data_structures/indexed_priority_queue.cpp

```diff
@@ -0,0 +1,258 @@
+	     */
+		bool contains(T1 key)
+		{
+            if (pm.count(key)) {  // c++11
+                return true;
+            }
+            return false;
+		}
+
+		/**
+		 * @brief inserts a key-value pair into the heap
+		 * @param key the key to insert
+		 * @param value the key's associated priority
+	     */
+		void insert(T1 key, T2 value)
+		{
+			// add node to bottom right of heap
+			val[key] = value;
+			pm[key] = size;
+			im[size] = key;
+
+			//swim
+			swim(size);
+			size++;
+		}
+
+		/**
+		 * @brief peeks the front of the priority queue
+		 * @returns a key value pair of T1 and T2
+	     */
+		std::pair<T1,T2> peek()
+		{
+			std::pair<T1,T2> x = std::make_pair(im[0], val[im[0]]);
+			return x;
+		}
+
+		/**
+		 * @brief removes the front element of the priority queue
+		 * @returns a key value pair of the front element before popping
+	     */
+		std::pair<T1,T2> pop()
+		{
+			std::pair<T1,T2> x = std::make_pair(im[0], val[im[0]]);
+			remove(im[0]); // O(logn)
+			return x;
+		}
+
+		/**
+		 * @brief removes a given key of the priority queue
+		 * @param key the key to remove
+	     */
+		void remove(T1 key)
+		{
+			if(!contains(key)) return;
+
+			int pos = pm[key];
+			swap(pos,size-1);
+			size--;
+
+			val.erase(key);
+			pm.erase(key);
+			im.erase(size);
+
+			sink(pos);
+			swim(pos);
+		}
+
+	private:
+		
+		/**
+		 * @brief swaps two key value pairs in the heap
+		 * @param i the heap index of the first pair
+		 * @param j the heap index ot the second pair
+	     */
+		void swap(int i, int j)
+		{
+			pm[im[j]] = i;
+			pm[im[i]] = j;
+			T1 tmp = im[i];
+			im[i] = im[j];
+			im[j] = tmp;
+		}
+		
+		/**
+		 * @brief An algorithm to set a misplaced item in a heap
+		 * to its correct place by swapping with children.
+		 * @param node the heap index to sink
+	     */
+		void sink(int node) 
+		{
+			while(true)
+			{
+				// children of a heap are represented as such
+				int left = (2*node)+1; 
+				int right = (2*node)+2;
+                int smallest = 0;
+
+                if (left >= size) {
+                    smallest = right;
+                } else if (right >= size) {
+                    smallest = left;
+                } else {
+                    smallest = (val[im[left]] < val[im[right]]) ? left : right;
+                }
+
+                if (left >= size && right >= size) {
+                    break;
+                }
+
+                if (val[im[smallest]] >= val[im[node]]) {
+                    break;
+                }
+
+                swap(smallest,node);
+				node = smallest;
+
+			}
+		}
+
+		/**
+		 * @brief An algorithm to set a misplaced item in a heap
+		 * to its correct place by swapping with parents.
+		 * @param node the heap index to swim
+	     */
+		void swim(int node) 
+		{
+			int i = std::floor((node-1)/2);
+			while(i >= 0 && val[im[node]] < val[im[i]])
+			{
+				swap(i,node);
+				node = i;
+				i = std::floor((node-1)/2);
+			}
+		}
+};
+
+static void test() {

```

Comment:
Still not solved and should be fixed.
```suggestion
/**
 * @brief Self-test implementations
 * @returns void
 */
static void test() {
```

Reply:
Apologies, added a description!